src/util/README.md

```diff
@@ -12,17 +12,57 @@ CPROVER codebase.
 
 \subsection irept_section irept: a 4-tuple (data, named-sub, comments, sub)
 
-See documentation at \ref irept.
+See detailed documentation at \ref irept.
 
-As that documentation says, [irept](\ref irept)s are generic tree nodes. You
-should think of them as having a single string ([data](irept::data), actually
-an \ref irep_idt) and lots of child nodes, some of which are numbered
+[irept](\ref irept)s are generic tree nodes. You
+should think of each node as holding a single string ([data](irept::data),
+actually an \ref irep_idt) and lots of child nodes, some of which are numbered
 ([sub](\ref irept::dt::sub)) and some of which are labelled, and the label
 can either start with a “\#” ([comments](\ref irept::dt::comments)) or without
 one ([named_sub](\ref irept::dt::named_sub)). The meaning of the “\#” is that
 this child should not be considered important, for example it shouldn't be
 counted when comparing two [irept](\ref irept)s for equality.
 
+They are used to represent all manner of structured objects throughout the
+CPROVER codebase, such as expressions, types and code. An \ref exprt represents
+expressions, including for example \ref equal_exprt, an equality predicate, or
+\ref dereference_exprt, which represents the `*` unary dereference operator
+found in C. A \ref typet represents a type, and may have other \ref typet nodes
+as children: for example, \ref array_typet represents an array, and has a single
+numbered child that gives the type of the array elements. Finally, \ref codet
+represents imperative code, such as \ref code_assignt, which represents an
```

Comment:
Is "statement" a term limited to imperative programming? Otherwise that's the term I'd use in place of "imperative code."

Reply:
Yes, but I'd still want to use the word "imperative" to make clear that's what `codet` has but `exprt` does not, so I'll leave this alone.

Reply:
I'm confused. The key difference is expressions vs. statements (programming-language terminology, not CPROVER terminology). An expression is implemented via `exprt`, a statement is implemented via `codet`. I don't see why "imperative" fits in here, unless, e.g., functional programming does not have a concept of a statement.

Reply:
If I said "statement" I'd want to clarify the difference between a statement in an imperative programming language (which codet essentially is, being stripped-down C) vs. e.g. a statement of truth, or assertion, so I'd write "imperative statement".

Reply:
Can we do something like "statement in source languages?" Maybe some paragraph to make clear that we are talking programming languages is required.

Reply:
Updated to "imperative statements," but I'm not going for "statement in source languages" since it is also used to represent statements in GOTO